### ISSUE TITLE
Cloudinary API POST Working on iOS but not Android

### ISSUE THREAD

**Issue by amblemind** [thunkable/thunkable-issues#562]
### Platform
✕ Android companion/downloaded - does not work
✕ iOS companion/downloaded - works

### Steps to Reproduce
1. Access my app - see below
2. I've stripped out everything not necessary to the app, so the app should open to the camera, so take a photo.

Cases:
1. If you are on iOS and your picture has text, you should see the text on the screen once the processing is done. 
2. If you are on iOS and you picture did NOT contain text, you should see an alert "no text found".
3. If you are on Android, you should get a 401 error from Cloudinary for invalid credentials. 

### Project Link
https://x.thunkable.com/projectPage/5ee305e02618e2327c6f21d5

This is a private client project that utilizes Cloudinary OCR. If a Thunkable staffer needs to remix this, please send me a message in the community. I am limited on OCR processing, so I can't have a ton of people testing this. 

### Expected Behavior
On android, I should see the same results as iOS. See steps to reproduce. 

### Actual Behavior
Android throws a 401 error for invalid credentials.

**Comment by Heikkiss** [comment]
I had the exact same problems, generally speaking, than the previous writer. Was trying to use Web API delete call to delete picture from Cloudinary -- it did work with iOS but not with Android, in any form. Such a shame as this would be a really sweet feature, and it already partially even works. Hope you could help us out here. Naturally I was using the Thunkable X.

Thank you!

Best, Heikki

**Comment by josmas** [comment]
Thanks for the report @amblemind and @Heikkiss 
We will look into it.

**Comment by Heikkiss** [comment]
Thank you @josmas !

**Comment by weihuali0509** [comment]
Hi - I created a simple app which open the photo library and upload to Cloudinary. It is working for me. Can to try if that work for you?

cc @josmas @TingC

**Comment by amblemind** [comment]
Not sure who that was intended for but...

Yes in the project linked above, uploading via the MediaDB works: 
![image](https://user-images.githubusercontent.com/45864164/84710501-f0ff3b00-af29-11ea-873e-3d72b0b54332.png)

Its the second step of POSTing back to that photo via the API that is not working. 
![image](https://user-images.githubusercontent.com/45864164/84710597-273cba80-af2a-11ea-86e9-9506eec2b71e.png)

**Comment by josmas** [comment]
@amblemind are you firing an authenticated POST request? According to the docs [here](
https://cloudinary.com/documentation/upload_images#uploading_with_a_direct_call_to_the_rest_api) you need a signature, and I cannot find one in your blocks; did I miss it?

As @weihuali0509 mentioned, the component works fine and it actually handles the signature for you alongside the type of resource depending on what `media` is. Am I guessing correctly in saying that you cannot use the component because you are making a call to the OCR API and not just uploading an asset?

**Comment by Heikkiss** [comment]
Hi all,

I've perhaps lost track a little of what you guys are discussing here but my problem was, simply put, that calling Cloudinary with the Web API component with the following structure does work with iOS but not with Android:

https://<API_KEY>:<API_SECRET>@api.cloudinary.com/v1_1/<CLOUD_NAME>/resources/image/upload?public_ids=<PUBLIC_ID>

As this is working with iOS it's a little puzzling why it doesn't work with Android. I've also tested it with Api tester and it works fine. With Android I get a response pointing to an authentication error. I'd need to have this to delete my pictures from the cloud as the Media DB component only lets me to upload pictures.

Sorry if this ventures a little off from your discussion but I just wanted to state my problem clearly. Would be nice if you could come up with anything to fix this.

Thank you!

Best, Heikki

**Comment by amblemind** [comment]
@josmas No. The photo is already uploaded. And I am using the Media DB component to upload the photo.

The issue for me is using this API request: https://cloudinary.com/documentation/image_upload_api_reference#upload_method

The bug is it works on iOS, but not on Android.

You can test my original project linked in the original post on both iOS and Android to see the issue with Android.

**Comment by weihuali0509** [comment]
Can someone create a simple project AND apk to demo this issue?

Thanks,
Wei

**Comment by josmas** [comment]
@amblemind so is that an unauthorized call? look at the optional parameters in that documentation link, are you using anything like `access_control` or `access_mode` or have you configured your account to receive unauthorized uploads? Otherwise the docs I linked seem to require a `signature`.

@Heikkiss I am also very puzzled that it works on iOS and not on Android; as per the docs [here](https://cloudinary.com/documentation/upload_images#uploading_with_a_direct_call_to_the_rest_api) it should **not work at all** without a `signature` unless you have changed your account's `cloud name` to allow unauthorized updates.

The docs are quite confusing, to be honest, I guess they just prefer people to use the libraries instead of direct web calls.

**Comment by amblemind** [comment]
@weihuali0509 Here are video demos of the app I linked in my original post. That app has everything stripped out. So it's about as simple as my use case is going to get. 
- iOS Demo: https://drive.google.com/file/d/1FgOiAUuwvK66tfXtcGE-ygigF3SQYNBj/view?usp=sharing
- Android Demo: https://drive.google.com/file/d/1KxW4TisHtJqldLF_7VZDUz6o64sTMCDl/view?usp=sharing

@Heikkiss can you share a demo of you DELETE call not working?

@josmas 
> The docs are quite confusing, to be honest, I guess they just prefer people to use the libraries instead of direct web calls.
Yes 100% agree with your observations. I'll look into creating a signature you linked above.

**Comment by Heikkiss** [comment]
Hi all,

So I didn't record a video but took a couple of screenshots. For this purpose here, I created a very simple demo app, which has a button that should do the deleting. After the delete (or an attempt at that) the app will show an alert with the API response.

So I took screenshots with Android and iOS (before and after hitting the button for both) and also of the blocks used. In this demo, I just defined a preset picture from Cloudinary, which I was able to delete with iOS but not with Android, which I tried first.

Hope this helps. It's really great that you're helping us. :)

By the way, sorry for my inactivity amid your discussion but it's also due to the time difference as I'm situated in Finland.

Best, Heikki

![Android_1](https://user-images.githubusercontent.com/60916326/84876493-f2e30e80-b08f-11ea-9296-5cfcbc63c96b.jpg)

![Android_2](https://user-images.githubusercontent.com/60916326/84876507-f6769580-b08f-11ea-8ca5-eaee09dabb03.jpg)

![iOS_1](https://user-images.githubusercontent.com/60916326/84876520-fa0a1c80-b08f-11ea-9268-5824044d8c50.png)

![iOS_2](https://user-images.githubusercontent.com/60916326/84876535-fd050d00-b08f-11ea-9577-e1ee386e70ce.png)

![Blocks](https://user-images.githubusercontent.com/60916326/84876547-00989400-b090-11ea-80b5-9bfcc156726f.jpg)

**Comment by amblemind** [comment]
So I was able to find another way to get my OCR feature working on Android. 

I used the upload API which uploads the photo a 2nd time with an upload_preset that returns the ocr text.

API: 
https://cloudinary.com/documentation/image_upload_api_reference#unsigned_upload_syntax

Preset: 
![image](https://user-images.githubusercontent.com/45864164/84928632-5bcc8580-b094-11ea-8529-3914b2748f03.png)

Blocks: 
![image](https://user-images.githubusercontent.com/45864164/84928700-730b7300-b094-11ea-9440-26ce411f2afd.png)
![image](https://user-images.githubusercontent.com/45864164/84928748-8585ac80-b094-11ea-8387-7d8516378c1f.png)

However, now I definitely need to get the delete photo api working. I'm seeing the same as @Heikkiss 

Doing a DELETE to this url 
`https://<API_KEY>:<API_SECRET>@api.cloudinary.com/v1_1/<CLOUD_NAME>/resources/image/upload?public_ids=<PUBLIC_ID> `
works on iOS. It also works when I plug it into other API testers like Postman. But it doesn't work in the Android app.

**Comment by jaredgib** [comment]
If it works in postman it should work on Thunkable. Do you have your
headers set correctly?

**Comment by amblemind** [comment]
@jaredgib I do not have any headers set in either.

**Comment by jaredgib** [comment]
When using POST and PUT calls to my personal api, I am required to send a
header of “Content-Type: application/json”

Weirdly, this is not the header that postman sends. Postman sends the
header “Content-Type: x-www-form-url-encoded”(or something close to that)

I wonder if these cloudinary calls need a header like that

**Comment by Heikkiss** [comment]
I have absolutely no clue as to how to proceed here but it's great to see that you're all working on this. Thanks again. Maybe we'll get there soon. I also wonder what causes the underlying disparity between Android and iOS--could this explain something? Naturally you need the Thunkable source code to check this.

**Comment by amblemind** [comment]
@jaredgib I tried removing the default headers in postman and was not able to replicate the response. I also tried adding headers in Thunkable and did not prove successful. 

However, I was able to delete a photo by doing the following. In my upload_preset detailed above, I turned on "return_delete_token". And then I used [this api](https://cloudinary.com/documentation/upload_images#deleting_client_side_uploaded_assets) call to delete it: 
`DELETE https://api.cloudinary.com/v1_1/<CLOUD_NAME>/delete_by_token?api_key=<API_KEY>&token=<RETURNED_DELETE_TOKEN>`

However, this delete token only lives for 10 minutes. This works for when I am only uploading a photo to read the text and then it can be deleted.

I'm noticing a trend in what works and doesn't work. 

**I think I can confidently say calling `https://api.cloudinary.com/v1_1/` apis work, but calling the "authenticated" version of this `https://<API_KEY>:<API_SECRET>@api.cloudinary.com/v1_1/` does not work on Android.**

**Comment by josmas** [comment]
All, I've looked into this and unfortunately I do not have an answer as for why Android and iOS behave differently. We only have one source for the WebAPI component, we use `fetch` and there are no branches like `if OS==Android do X else if OS==iOS do Y" so it is the same code.

I've compared for responses and the only thing I can see if that Android responds with the following:
```
"status": "401 Unauthorized",
"www-authenticate": "Basic realm=\"Cloudinary\"",
"x-cld-error": "Invalid credentials",
[...]
"ok": false,
"status": 401,
```

while iOS responds with:
```
"status": "200 OK",
"vary": "Accept-Encoding",
[...]
"ok": true,
"status": 200,
```

So the difference has to be either on their side, or a particular undocumented header they have not made explicit in their docs for Android to place the right request. `Curl` gets the same results as iOS. So other than opening an issue with them, not sure where to go from here.

**Comment by josmas** [comment]
I have opened an [issue](https://github.com/cloudinary/cloudinary-react/issues/165) with cloudinary.

**Comment by Heikkiss** [comment]
This is fantastic. Thank you @josmas ! Hope it will be resolved. If so, would you mind updating here?

Best, Heikki

**Comment by josmas** [comment]
Hey @Heikkiss I will certainly keep an eye on the issue; so far no answer but we'll see. Thanks!

**Comment by Heikkiss** [comment]
Thank you @josmas !

**Comment by josmas** [comment]
FYI: they've asked me to open a support request. I'll let you all know when I hear back.

**Comment by Heikkiss** [comment]
Great! Good news. Thanks @josmas

**Comment by josmas** [comment]
Hi all, so I got a response from Cloudinary that may work for some of you. The idea is to add an authorization Header for Android, as that platform will not allow receiving any KEY information in the URL. So the trick is to use this URL and also add the block below:

URL: `https://api.cloudinary.com/v1_1/<cloud name>/resources/image/upload?public_ids=<public_id of the resource>`

<img width="788" alt="Screen Shot 2020-07-08 at 5 22 54 PM" src="https://user-images.githubusercontent.com/194925/86944782-228daf80-c140-11ea-9a3c-ab8dc33e6614.png">

Note that the token is `base64` encoded, and it is a combination of your keys in the same way you were passing them in the other URL `FIRST_KEY:SECOND_KEY`, with a colon in between them. Unfortunately we do not have a block to encode in base64 at the moment (as far as I know).

In a platform like nodejs you can do this with a script such as:
```
let data = 'API_KEY:API_SECRET';
let buff = new Buffer(data);
let base64data = buff.toString('base64');

console.log('"' + data + '" converted to Base64 is "' + base64data + '"');
```

So the idea here is that you place whatever gets printed out as `base64data` after the word `Basic`. Note that capitals in the header are very important in both parts of the JSON object.

There are also a number of online sites that can do the base64 encoding for you such as https://www.base64encode.org/

I have tried this and I get a 200 back from the server. Let me know if that works if you get a chance to test it cc/ @Heikkiss @amblemind

**Comment by Heikkiss** [comment]
Hi @josmas -- and many, many thanks for finding this out. I'm currently on holiday and have a limited access to computer, but I'll look into this once I return to normal life... :-) In any case, this looks really promising and big thanks to you for seeking for solution.

Best, Heikki

**Comment by Heikkiss** [comment]
Hi there again @josmas & @amblemind 

So, I finally got back to regular life and tried out your advice immediately. First I felt like this was beyond my competence level (not really a programmer or anything) but I just followed your advice according to my understanding of it and ... it worked! :) Wow! So, to try this out, I modified my test app (with a button to delete a preset picture from the cloud).

Nicely done, and many many thanks for this @josmas ! Of course for the future it might be a good idea to include this functionality as a more easily accessible feature in Thunkable (as deleting from cloud is quite commonplace I'm sure), but at least we got it working this way anyway.

I'll also attach here a screenshot of what I got after the delete was done.

Thank you again! Well done! It's so nice to get such a good support with Thunkable.

![Screenshot_20200803_160649_com thunkable live](https://user-images.githubusercontent.com/60916326/89186324-4bb92880-d5a4-11ea-9385-a602cc9b9790.jpg)
